src/common/models/data-series/data-series.ts

```diff
@@ -0,0 +1,143 @@
+  format: SeriesFormat;
+}
+
+const defaultDataSeries: DataSeriesValue = {
+  measure: null,
+  percentOf: null,
+  format: DEFAULT_FORMAT
+};
+
+export enum DataSeriesPercentOf { PARENT = "of_parent", TOTAL = "of_total" }
+
+export interface Period {
+  derivation: SeriesDerivation;
+  filter: Expression;
+}
+
+export class PreviousPeriod implements Period {
+  derivation = SeriesDerivation.PREVIOUS;
+
+  constructor(public filter: Expression) {
+  }
+}
+
+export class CurrentPeriod implements Period {
+  derivation = SeriesDerivation.CURRENT;
+
+  constructor(public filter: Expression) {
+  }
+}
+
+export class DataSeries extends Record<DataSeriesValue>(defaultDataSeries) {
+
+  public fullName(derivation = SeriesDerivation.CURRENT): string {
+    return fullName(this.measure.name, derivation, this.percentOf);
+  }
+
+  private filterMainRefs(exp: Expression, filter: Expression): Expression {
+    return exp.substitute(e => {
+      if (e instanceof RefExpression && e.name === "main") {
+        return $("main").filter(filter);
+      }
+      return null;
+    });
+  }
+
+  private applyPeriod(period: Period): Expression {
+    if (!period) return this.measure.expression;
+    return this.filterMainRefs(this.measure.expression, period.filter);
+  }
+
+  private relativeNesting(nestingLevel: number): number {
+    switch (this.percentOf) {
+      case DataSeriesPercentOf.TOTAL:
+        return nestingLevel;
+      case DataSeriesPercentOf.PARENT:
+        return Math.min(nestingLevel, 1);
+    }
+  }
+
+  private percentOfExpression(expression: Expression, nestingLevel: number, derivation = SeriesDerivation.CURRENT): ApplyExpression {
+    const name = this.fullName(derivation);
+    const formulaName = `__formula_${name}`;
+    if (nestingLevel > 0) {
+      return new ApplyExpression({
+        name,
+        operand: new ApplyExpression({ expression, name: formulaName }),
+        expression: $(formulaName).divide($(formulaName, nestingLevel))
+      });
+    }
+    if (nestingLevel === 0) {
+      return new ApplyExpression({ name: formulaName, expression });
+    }
+    throw new Error(`wrong nesting level: ${nestingLevel}`);
+  }
+
+  private toApplyExpression(expression: Expression, currentNesting: number, derivation: SeriesDerivation): ApplyExpression {
+    if (!this.percentOf) return new ApplyExpression({ name: this.fullName(derivation), expression });
+    return this.percentOfExpression(expression, this.relativeNesting(currentNesting), derivation);
+  }
+
+  public toExpression(nestingLevel: number, period?: Period): ApplyExpression {
+    const expWithPeriodFilter = this.applyPeriod(period);
+    return this.toApplyExpression(expWithPeriodFilter, nestingLevel, period && period.derivation);
+  }
+
+  public title(derivation = SeriesDerivation.CURRENT): string {
+    return title(this.measure.title, derivation, this.percentOf);
+  }
+
+  public getDatum(datum: Datum, derivation = SeriesDerivation.CURRENT) {
```

Comment:
I don't really like this kind of delegation. It would be more explicit with Datum method like datum.resolve(dataSeries.something). Could you also check dependences between datum and dataseries? We should avoid cycles.

Reply:
Ok, it's poorly named, it's something more like "getDataFromDatum". 

Datum is just a POJO returned from plywood. It's just a Map from measure names to values. DataSeries represents Series of data visible on Viz. Because DataSeries was used to generate query it also knows how to get data from Datum with proper key.

Idea is to give components DataSeries object, so component could be oblivious to essence->plywood->datum flow. This is a description how to get data, how to get title, and how to format data. 

I admit, there is indirection, but on call site, in components it simplify code.